src/scores/processing/block_bootstrap.py

```diff
@@ -0,0 +1,410 @@
+        n_iteration: The number of iterations to repeat the bootstrapping process. Determines
+            how many bootstrapped arrays will be generated and stacked along the iteration
+            dimension.
+        exclude_dims: An optional parameter indicating the dimensions to be excluded during
+            bootstrapping for each object provided in `objects`. This parameter expects a list
+            of lists, where each inner list corresponds to the dimensions to be excluded for
+            the respective object. By default, the assumption is that no dimensions are
+            excluded, and all objects are bootstrapped across all specified dimensions in `blocks`.
+        circular: A boolean flag indicating whether circular block bootstrapping should be
+            performed. Circular bootstrapping means that bootstrapping continues from the beginning
+            when the end of the data is reached. By default, this parameter is set to True.
+
+     Returns:
+        Tuple of bootstrapped xarray DataArrays or Datasets, based on the input.
+
+    Note:
+        This function expands out the iteration dimension inside a universal function.
+        However, this may generate very large chunks (multiplying chunk size by the number
+        of iterations), causing issues for larger iterations. It's advisable to apply this
+        function in blocks using 'block_bootstrap'.
+
+    References:
+    Wilks, Daniel S. Statistical methods in the atmospheric sciences. Vol. 100.
+      Academic press, 2011.
+    """
+
+    objects_list = list(objects)
+
+    # Rename exclude_dims so they are not bootstrapped
+    if exclude_dims is None:
+        exclude_dims = [[] for _ in range(len(objects_list))]
+    if (
+        not isinstance(exclude_dims, list)
+        or len(exclude_dims) != len(objects_list)
+        or not all(isinstance(x, list) for x in exclude_dims)
+    ):
+        raise ValueError(
+            "exclude_dims should be a list of the same length as the number of "
+            "objects containing lists of dimensions to exclude for each object"
+        )
+    renames = []
+    for i, (obj, exclude) in enumerate(zip(objects_list, exclude_dims)):
+        objects_list[i] = obj.rename(
+            {d: f"dim{ii}" for ii, d in enumerate(exclude)},
+        )
+        renames.append({f"dim{ii}": d for ii, d in enumerate(exclude)})
+
+    dim = list(blocks.keys())
+
+    # Ensure bootstrapped dimensions have consistent sizes across objects_list
+    for d in blocks.keys():
+        dim_sizes = [o.sizes[d] for o in objects_list if d in o.dims]
+        if not all(s == dim_sizes[0] for s in dim_sizes):
+            raise ValueError(f"Block dimension {d} is not the same size on all input objects")
+
+    # Retrieve sizes of the bootstrap dimensions from objects_list
+    sizes = None
+    for obj in objects_list:
+        try:
+            sizes = OrderedDict(
+                {d: (obj.sizes[d], b) for d, b in blocks.items()},
```

Comment:
I believe you have to initialize an ordered `dict` as a list of tuples, otherwise it doesn't maintain order on initialization. In more recent versions of python `dict` is already ordered I believe, so this behaviour may have been obfuscated.

This is a safer way, since lists by definition are ordered (assuming the way I remember initializing ordered dicts is still accurate):

```python
sizes = OrderedDict([
    (d, (obj.sizes[d], b))
    for d,b in blocks.items()
])
```